motorutil/steer.go

```diff
@@ -0,0 +1,276 @@
+// Copyright ©2016 The ev3go Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package motorutil
+
+import (
+	"fmt"
+	"math"
+	"sync"
+	"time"
+
+	"github.com/ev3go/ev3dev"
+)
+
+// Steering implements a paired-motor steering unit similar to an EV-G steering block.
+type Steering struct {
+	// Left and Right are the left and right motors to be
+	// used by the steering module.
+	Left, Right *ev3dev.TachoMotor
+
+	// Timeout is the timeout for waiting for motors to
+	// return to a non-driving state.
+	//
+	// See ev3dev.Wait documentation for timeout behaviour.
+	Timeout time.Duration
+}
+
+// SteerCounts steers in the given direction for the given tacho counts and at the
+// specified speed. The valid range of direction is -1 (hard left) to +1 (hard right).
+// For directions between -1 and 1 neither motor will run backwards, but for -1 and 1
+// the motors will run in opposite directions to allow a turn on the spot in the given
+// direction.
```

Comment:
In the EV3-G software, the range is -100 to 100 for steering. For these values, the robot will basically turn in place (left runs forward and right runs backwards at same speed).

I think this is what you are getting at, but saying "neither motor will run backwards" doesn't sound right.

Reply:
Another thing, the EV3 does not have a floating point unit, so anything that uses floats will be less efficient. I don't know that there would be a significant (noticeable) difference here. It is just something to consider. I generally try to avoid float just on principal.

Reply:
Another more useful thing to think about is one feature of the official LEGO firmware that is frequently requested in ev3dev is that motors are "synchronized", meaning that if one has a higher load than the other, the faster one will wait for the other to catch up, so to speak. I think the leJOS guys have a blog about this for an arbitrary number of motors. LEGO does this in the kernel drivers for 2 motors, but ev3dev doesn't to it at all, so it would have to be implemented in Go.

Reply:
>I think this is what you are getting at, but saying "neither motor will run backwards" doesn't sound right.

When I've experimented, there is forward progression except with very tight turns, where you see a complete spin. The comment about no reverse motor for the other cases is true (at least for this code).

>Another thing, the EV3 does not have a floating point unit, so anything that uses floats will be less efficient. I don't know that there would be a significant (noticeable) difference here. It is just something to consider. I generally try to avoid float just on principal.

I don't think there is a huge issue here since it's only a very minor part of the code path.

I'll look into synchronising the motors, but it's not a juge priority.

Reply:
From looking at the leJOS article, it's not clear that I'd be able to do very much better than what is there now without significant added complexity and running the motors with `run-direct`. Fortunately, because the Go code is natively compiled there is no JIT warmup.

Reply:
I have changed `direction` to take an `int` argument; it simplifies the API and removes the need for a new error interface type.

Reply:
You're right about the turns; I can't read my own code which is upsetting. I have clarified the docs and added tests to demonstrate the expected behaviour.

[This](http://stemrobotics.cs.pdx.edu/sites/default/files/EV3%20Move%20Steering%20Block%20Explained.pdf) is useful for future reference.